lucene/core/src/java/org/apache/lucene/search/TopDocsCollector.java

```diff
@@ -136,14 +136,20 @@ public TopDocs topDocs(int start, int howMany) {
     // pq.size() or totalHits.
     int size = topDocsSize();
 
-    // Don't bother to throw an exception, just return an empty TopDocs in case
-    // the parameters are invalid or out of range.
-    // TODO: shouldn't we throw IAE if apps give bad params here so they dont
-    // have sneaky silent bugs?
-    if (start < 0 || start >= size || howMany <= 0) {
+
+    if (start < 0 || start > howMany) {
+      throw new IllegalArgumentException("Expected value of starting position is between 0 and " + size +
+          ", got " + howMany);
+    }
+
+    if (start >= size) {
       return newTopDocs(null, start);
     }
 
+    if (howMany <= 0) {
+      throw new IllegalArgumentException("Number of hits requested must be greater than 0 but value was " + howMany);
+    }
```

Comment:
can you move this up? We should validate input parameters before we do anything else?